src/core/async-work-cache.ts

```diff
@@ -0,0 +1,68 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+/**
+ * A map from keys to promises of values. Used for caching asynchronous work.
+ */
+export class AsyncWorkCache<K, V> {
+  private _keyToResultMap: Map<K, Promise<V>>;
```

Comment:
I can see that. There are some fuzzy bits that I want to double check, but I left a TODO.